### PR TITLE
net: l2: ieee802154: unified timestamp preparation

### PR DESCRIPTION
**Pull request by ghost**
This pull request works towards ongoing realization of the following RFCs:

* #50336
* Layer 1 (clock sources) and layer 2 (syntonization) of a [high-level clock subsystem](https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1597226731)
* [Improved cross-vendor support of the IEEE 802.15.4 driver and L2 stack](https://github.com/zephyrproject-rtos/zephyr/pull/58438#discussion_r1220059230)

~~The kernel part of this PR needs to be approved in a separate PR (#60400) before this PR can be merged. It is provided as an example application of ktime_t and for preliminary preview wrt the above cited RFCs.~~

UPDATE: It has been proposed by myself and @cfriedt to decouple the two PRs as the kernel part needs more work and shouldn't block the unification of timestamps in the network subsystem - which is useful on its own.

This PR is a precondition for fixing #59245.

### CODE REVIEW

**Comment by fgrandel** [comment]
This PR can now be reviewed. It has been proposed in #60400 to not couple the development of the larger clock subsystem to this change as this change is useful in itself and shouldn't be held back due to other PRs depending on it.

@nandojve @jukkar Probably also interesting to you as this introduces a unified concept for all timestamps and time spans in the network stack.

**Comment by fgrandel** [comment]
@edmont Heads up to you as this is the change set that prepares for the fix of #59245 which we're currently working on together.

**Comment by fgrandel** [comment]
@edmont @rlubos @jukkar I'm not very happy with the ktime_t name - it was named like in Linux as it is the same concept and was meant to be part of a larger clock subsystem but that will take time. I like the idea to have a common well-specified type for scalar nanosecond timestamps in the network subsystem, though, and I believe that this PR shows that the idea works rather nicely. Maybe you have a better idea how to name this type? It should still be possible to rename it later when it is replaced by something more generic.

net_timestamp_t? net_ns_uptime_t?

**Review by rlubos — COMMENTED** [review]
`net_timestamp_t` sounds ok.

File: include/zephyr/net/net_ktime.h [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+ * applications. Usage of this file is currently restricted to representation of
+ * nanosecond resolution scalar time and time spans within network subsystem.
+ *
+ * TODO: Either make these definitions public if they turn out to be generally
+ * useful in a larger Zephyr clock subsystem or change the name of the type to
+ * something network specific if the general clock subsystem will not use this
+ * type, see the ongoing discussion in #60400 for details.
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_KTIME_H_
```
[the remark this answers is not in the document]
Comment (fgrandel) [reply include/zephyr/net/net_ktime.h:26]: Fixed.

File: include/zephyr/net/ieee802154_radio.h [hunk 2]
```diff
@@ -382,77 +466,251 @@ struct ieee802154_config {
...
+	/**
+	 * @brief Set/Unset PAN ID, extended or short address filters.
+	 *
+	 * @note requires IEEE802154_HW_FILTER capability.
+	 *
+	 * @param dev pointer to radio device
+	 * @param set true to set the filter, false to remove it
+	 * @param type the type of entity to be added/removed from the filter
+	 * list (a PAN ID or a source/destination address)
+	 * @param type the entity to be added/removed from the filter
```
[the remark this answers is not in the document]
Comment (fgrandel) [reply include/zephyr/net/ieee802154_radio.h]: Fixed.

**Comment by fgrandel** [comment]
> `net_timestamp_t` sounds ok.

Renamed to net_time_t as you know - so that it also intuitively represent time spans. The result now feels much more natural to me, thanks for you feedback there.

**Review by rlubos — DISMISSED** [review]

**Review by edmont — APPROVED** [review]

**Review by jukkar — COMMENTED** [review]
I cannot do now full review, but I faintly recall that the timestamp and txtime fields in net_pkt were suppose to be separate so that both can be enabled / used at the same time. Did you check this when doing the pr?

**Comment by fgrandel** [comment]
> I cannot do now full review, but I faintly recall that the timestamp and txtime fields in net_pkt were suppose to be separate so that both can be enabled / used at the same time. Did you check this when doing the pr?

@jukkar Yes I did and I didn't find any use case where these would be used together or even where it theoretically makes sense to use them together. It also has to be mentioned that Linux has only one field as well if I'm not mistaken.

But of course I could have overlooked something. So it would be great if this result could be confirmed by others.

**Comment by fgrandel** [comment]
@jukkar Oh and btw - it's not a problem if this PR is not merged quickly. I prefer thorough review over fast merge. :-) So if it's just a matter of doing review later, that is of course totally fine. You review would be very much appreciated.

**Review by jukkar — COMMENTED** [review]
Ok, that is good news. I am currently having a vacation and travelling but I am trying to review the pr asap.

**Comment by fgrandel** [comment]
@rlubos @edmont @jukkar The recently merged get_attr method caused merge conflicts which are now resolved. I also added more documentation to that new attribute feature, especially focusing on potential abuse. @tbursztyka This hopefully addresses the concerns that you had against introducing a getter. Please note that this is NOT a configuration getter but reserved to scalar/structured attributes that can neither be modeled as boolean HW CAPs nor as L2-driven configuration but originate from the driver itself (e.g. some of the PHY PIB attributes/constants).

**Review by jukkar — DISMISSED** [review]
LGTM. Very nice to see extensive documentation of various structs. I think we need to add something to release notes as some api changes are there.

**Comment by fgrandel** [comment]
@jukkar Thanks for your review! Release notes added.

@jukkar @rlubos Think this is now ready for final approval and merge.

**Review by rlubos — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]
🪁

**Review by jukkar — APPROVED** [review]

**Pull request merged**